peer.go

```diff
@@ -645,13 +654,20 @@ func (p *peer) Disconnect(reason error) {
 	peerLog.Infof("Disconnecting %s, reason: %v", p, reason)
 
 	// Ensure that the TCP connection is properly closed before continuing.
-	p.conn.Close()
+	// Connection may be nil in tests.
+	if p.conn != nil {
+		p.conn.Close()
+	}
 
 	close(p.quit)
 }
 
 // String returns the string representation of this peer.
 func (p *peer) String() string {
+	if p.conn == nil {
```

Comment:
Ideally we should set it in the test instead, to avoid special casing this.

Reply:
would agree if it doesn't add too much code, otherwise we can remove this when we expand our mock connections and use a proper interface inside the peer